[visionOS] WebViews need to request the "Two Handed Interaction Processor" behavior

etiennesegonzac opened the pull request:
#### 967b8318789f7e0dd92f68fd31d944f0609593db
<pre>
[visionOS] WebViews need to request the &quot;Two Handed Interaction Processor&quot; behavior
<a href="https://bugs.webkit.org/show_bug.cgi?id=299118">https://bugs.webkit.org/show_bug.cgi?id=299118</a>
&lt;<a href="https://rdar.apple.com/160683670">rdar://160683670</a>&gt;

Reviewed by Abrar Rahman Protyasha.

When available, add the RemoteEffect that lets us opt-into the correct
behavior for two handed gestures on the top level scroll view.

* Source/WebCore/PAL/pal/spi/cocoa/QuartzCoreSPI.h:
Forward-declaration.
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
Add missing public SDK stub.
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _setupScrollAndContentViews]):
Add the effect when available.

Canonical link: <a href="https://commits.webkit.org/300291@main">https://commits.webkit.org/300291@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8343356bed6f30941752c11f2da364d8c4833e5c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122017 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41719 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32389 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128580 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74110 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/8b9f002f-ab12-4246-88eb-24d55e4cd2cf) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/123893 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42434 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50313 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92734 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61619 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2cfa05c0-8fb3-4471-8179-565f0afa8719) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124969 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33838 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109269 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73388 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/bdf32180-35e4-428d-b2cc-83a5a0459dcc) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32851 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27434 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72074 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103345 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27626 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131341 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48956 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37229 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101294 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49330 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105483 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101165 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25655 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46536 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24653 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/45669 "Built successfully") | | 
| | [❌ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48813 "Hash 8343356b for PR 50952 does not build (failure)") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54547 "Built successfully") | | | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48283 "Hash 8343356b for PR 50952 does not build (failure)") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51633 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49963 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/PAL/pal/spi/cocoa/QuartzCoreSPI.h:
```diff
@@ -361,3 +361,7 @@ extern NSString * const kCASnapshotOriginY;
 extern NSString * const kCASnapshotTimeOffset;
 #endif
+
+@interface CARemoteExternalEffect (Radar_160881286)
```
You'll probably want to include `CARemoteEffectPrivate.h` for `USE(APPLE_INTERNAL_SDK)`, and declare `@interface CARemoteExternalEffect : CARemoteEffect;` for `!USE(APPLE_INTERNAL_SDK)` first.

aprotyas replied:
Also, maybe leave a FIXME to clean up further when the interface is available in more SDKs?

etiennesegonzac committed:
[visionOS] WebViews need to request the "Two Handed Interaction Processor" behavior
https://bugs.webkit.org/show_bug.cgi?id=299118
<rdar://160683670>

Reviewed by Abrar Rahman Protyasha.

When available, add the RemoteEffect that lets us opt-into the correct
behavior for two handed gestures on the top level scroll view.

* Source/WebCore/PAL/pal/spi/cocoa/QuartzCoreSPI.h:
Forward-declaration.
* Source/WebKit/Platform/spi/ios/UIKitSPI.h:
Add missing public SDK stub.
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _setupScrollAndContentViews]):
Add the effect when available.

Canonical link: https://commits.webkit.org/300291@main
